### ISSUE TITLE
How to finetune this model

### ISSUE THREAD

**Issue by yashDriveX** [Bria-AI/RMBG-2.0#3]

**Comment by YueQiang-ctw** [comment]
I also want to know it

**Comment by Algorithman** [comment]
You have your originals and your masks?
then it's quite easy.
I've done it with Rocm, Radeon 9070XT.
Had another dataset with ~5000 images, after 3 epochs output was quite nice, let it cook over night and it adjusted really good.

What i DON'T know is how to export this to onnx. I even got back to torch2.5.0 (their onnx was created with it), but I can't get it to trace/compile/convert the atrous_conv.

~~If anyone knows, please tell me.~~ See edit at the end.

```
import ctypes
import os
from tqdm import tqdm

import torch
import torch.nn as nn
from torch.utils.data import Dataset, DataLoader, random_split
from torchvision import transforms
from PIL import Image
from transformers import AutoModelForImageSegmentation

# ----------------------------
# CONFIG
# ----------------------------
IMG_DIR = r"your_input_folder_to\original"
MASK_DIR = r"your_input_folder_to\masks"

MAX_FILES = 6000
MODEL_PATH = r"your_path_to_the_model"
SAVE_PATH = r"E:\path_to\bria-finetuned"  # folder!

IMAGE_SIZE = (800, 800) # Adjust this for your VRAM!! higher values can crash you hard, 768 seems fine for a 16GB GPU
BATCH_SIZE = 1
EPOCHS = 50
LR = 1e-5

DEVICE = torch.device("cuda" if torch.cuda.is_available() else "cpu")


# ----------------------------
# DATASET
# ----------------------------
class SegDataset(Dataset):
    def __init__(self, img_dir, mask_dir, max_files=None):
        self.files = []
        for f in os.listdir(img_dir):
            if f.lower().endswith((".jpg", ".jpeg")):
                mask_path = os.path.join(mask_dir, os.path.splitext(f)[0] + ".png")
                if os.path.exists(mask_path):
                    self.files.append(f)

        import random
        random.seed(42)
        random.shuffle(self.files)

        if max_files is not None:
            self.files = self.files[:max_files]

        self.img_dir = img_dir
        self.mask_dir = mask_dir

        self.img_tf = transforms.Compose([
            transforms.Resize(IMAGE_SIZE),
            transforms.ToTensor(),
            transforms.Normalize([0.485, 0.456, 0.406],
                                 [0.229, 0.224, 0.225])
        ])
        self.mask_tf = transforms.Compose([
            transforms.Resize(IMAGE_SIZE),
            transforms.ToTensor()
        ])

    def __len__(self):
        return len(self.files)

    def __getitem__(self, idx):
        name = self.files[idx]
        img = Image.open(os.path.join(self.img_dir, name)).convert("RGB")
        mask = Image.open(
            os.path.join(self.mask_dir, os.path.splitext(name)[0] + ".png")
        ).convert("L")
        img = self.img_tf(img)
        mask = self.mask_tf(mask)
        ## mask = (mask > 0.5).float()
        return img, mask


# ----------------------------
# HELPERS
# ----------------------------
def permanently_freeze_bn(model: nn.Module):
    """Put every BN layer into eval mode and freeze its parameters."""
    for module in model.modules():
        if isinstance(module, (nn.BatchNorm1d, nn.BatchNorm2d, nn.BatchNorm3d,
                                nn.SyncBatchNorm)):
            module.eval()
            module.track_running_stats = True
            if module.weight is not None:
                module.weight.requires_grad = False
            if module.bias is not None:
                module.bias.requires_grad = False


bce = nn.BCEWithLogitsLoss()

def dice(pred, target, eps=1e-6):
    pred = torch.sigmoid(pred)
    inter = (pred * target).sum(dim=(2, 3))
    union = pred.sum(dim=(2, 3)) + target.sum(dim=(2, 3))
    return 1 - ((2 * inter + eps) / (union + eps)).mean()

def loss_fn(pred, target):
    return bce(pred, target) + dice(pred, target)

def forward(model, x):
    outputs = model(x)

    if isinstance(outputs, torch.Tensor):
        candidates = [outputs]
    elif isinstance(outputs, (list, tuple)):
        candidates = [o for o in outputs if torch.is_tensor(o)]
    elif isinstance(outputs, dict):
        candidates = [v for v in outputs.values() if torch.is_tensor(v)]
    else:
        candidates = []

    best, best_size = None, 0
    for c in candidates:
        if c.ndim == 4:
            spatial = c.shape[-1] * c.shape[-2]
            if spatial > best_size:
                best = c
                best_size = spatial

    if best is None:
        raise RuntimeError("No valid prediction tensor found in model output")
    return best


# ----------------------------
# MAIN
# ----------------------------
if __name__ == '__main__':

    # Load model
    model = AutoModelForImageSegmentation.from_pretrained(
        MODEL_PATH,
        trust_remote_code=True,
        local_files_only=True
    ).to(DEVICE)

    # Freeze all BN layers
    permanently_freeze_bn(model)

    # ✅ Patch .train() so BN stays frozen no matter what
    _original_train = model.__class__.train

    def _train_with_frozen_bn(self, mode=True):
        _original_train(self, mode)
        permanently_freeze_bn(self)
        return self

    model.__class__.train = _train_with_frozen_bn

    # Freeze encoder
    for name, p in model.named_parameters():
        if "encoder" in name:
            p.requires_grad = False

    # Freeze aux/deep supervision heads
    for name, p in model.named_parameters():
        if any(k in name for k in ("aux", "deep", "side")):
            p.requires_grad = False

    # Unfreeze decoder + prediction head
    for name, p in model.named_parameters():
        if any(k in name for k in ("decoder", "head")):
            p.requires_grad = True

    # Re-freeze BN params after requires_grad changes
    permanently_freeze_bn(model)

    trainable = sum(p.numel() for p in model.parameters() if p.requires_grad)
    print(f"Trainable parameters: {trainable:,}")

    # Data
    dataset = SegDataset(IMG_DIR, MASK_DIR, MAX_FILES)
    train_len = int(0.8 * len(dataset))
    val_len = len(dataset) - train_len
    train_ds, val_ds = random_split(dataset, [train_len, val_len])

    train_loader = DataLoader(train_ds, batch_size=BATCH_SIZE, shuffle=True,
                              num_workers=0, pin_memory=True)
    val_loader   = DataLoader(val_ds,   batch_size=BATCH_SIZE, shuffle=False,
                              num_workers=0, pin_memory=True)

    # Optimizer
    optimizer = torch.optim.AdamW(
        filter(lambda p: p.requires_grad, model.parameters()),
        lr=LR, weight_decay=1e-4
    )
    scaler = torch.cuda.amp.GradScaler()

    def validate(pbar=None):
        total = 0.0
        # Wrap val_loader with tqdm if pbar is provided, otherwise standard loop
        loader_iter = tqdm(val_loader, desc="Validating", leave=False, disable=pbar is not None)
        
        with torch.no_grad():
            for imgs, masks in loader_iter:
                imgs, masks = imgs.to(DEVICE), masks.to(DEVICE)
                with torch.cuda.amp.autocast():
                    preds = forward(model, imgs)
                    total += loss_fn(preds, masks).item()
        return total / len(val_loader)

    # Train
    best_val = float("inf")
    print("Training BiRefNet (decoder only, BN permanently frozen)...")

    for epoch in range(EPOCHS):
        train_loss = 0.0
        
        # Wrap train_loader with tqdm
        train_iter = tqdm(train_loader, desc=f"Epoch {epoch+1}/{EPOCHS}", leave=True)
        
        for i, (imgs, masks) in enumerate(train_iter):
            imgs, masks = imgs.to(DEVICE), masks.to(DEVICE)
            optimizer.zero_grad()

            with torch.cuda.amp.autocast():
                preds = forward(model, imgs)
                loss = loss_fn(preds, masks)

            scaler.scale(loss).backward()
            scaler.step(optimizer)
            scaler.update()

            train_loss += loss.item()
            
            # Update progress bar description with current batch loss
            train_iter.set_description(f"Epoch {epoch+1}/{EPOCHS} | Loss: {loss.item():.4f}")

        val_loss = validate()
        avg_train = train_loss / len(train_loader)
        print(f"\nEpoch {epoch+1} — Train: {avg_train:.4f} | Val: {val_loss:.4f}")

        if val_loss < best_val:
            best_val = val_loss
            model.save_pretrained(SAVE_PATH)
            print(f"  ✅ Saved best model (val={best_val:.4f})")

    print("Done.")
```

Have fun!


Edit: Just found the answer to my problem.
for onnx export refer to [Kazuhito00/BiRefNet-ONNX-Sample](https://github.com/Kazuhito00/BiRefNet-ONNX-Sample/blob/main/Convert2ONNX.ipynb).
Step 6 and 7 are important.

**Comment by YueQiang-ctw** [comment]
thanks @Algorithman ， you code help me a lot。
